File: src/tools/tests/test_ScopeSerial.m
%
% Copyright (c) 2021-present LAAS-CNRS
%
% SPDX-License-Identifier: GPL-2.0-or-later
%

classdef test_ScopeSerial < matlab.unittest.TestCase
    methods (Test)
        function explicitProbeAndCleanup(testCase)
            testCase.verifyError( ...
                @() ScopeSerial(""), "ScopeSerial:validation");

            transport = FakeScopeTransport( ...
                uint8.empty(1, 0), uint8(sprintf('WRONG PORT\n')));
            testCase.verifyError( ...
                @() ScopeSerial( ...
                    "fake", Transport=transport, ProbeTimeout=0.01), ...
                "ScopeSerial:timeout");
            testCase.verifyTrue(transport.Closed);
        end

        function validCaptureIsDecodedAndRotated(testCase)
            transport = FakeScopeTransport( ...
                test_ScopeSerial.makeRecord());
            scope = ScopeSerial("fake", Transport=transport);
            capture = scope.download(test_ScopeSerial.metadata(10, 0.2));

            testCase.verifyEqual(transport.Writes, ...
                {uint8('?'), uint8('D')});
            testCase.verifyEqual( ...
                capture.channelNames, ScopeSerial.ChannelNames);
            testCase.verifyEqual(capture.finalIndex, 1022);
            testCase.verifyEqual(capture.samples(1, 1), 10230);
            testCase.verifyEqual(capture.samples(2, 1), 0);
            testCase.verifySize(capture.samples, [1024, 8]);
            testCase.verifyEqual( ...
                capture.timeAxisS(1), -0.2048, AbsTol=1e-12);
            scope.close();
            testCase.verifyTrue(transport.Closed);
        end

        function malformedPayloadsAreRejected(testCase)
            records = {
                test_ScopeSerial.makeRecord(ValueCount=8191), ...
                test_ScopeSerial.makeRecord(CorruptValue=123), ...
                test_ScopeSerial.makeRecord(ValueCount=8193)
            };
            for i = 1:numel(records)
                transport = FakeScopeTransport(records{i});
                scope = ScopeSerial("fake", Transport=transport);
                testCase.verifyError( ...
                    @() scope.download(test_ScopeSerial.metadata(1, 0)), ...
                    "ScopeSerial:protocol");
                scope.close();
            end
        end

        function wrongChannelAndFinalIndexAreRejected(testCase)
            channels = ScopeSerial.ChannelNames;
            channels(1) = "wrong";
            transport = FakeScopeTransport( ...
                test_ScopeSerial.makeRecord(Channels=channels));
            scope = ScopeSerial("fake", Transport=transport);
            testCase.verifyError( ...
                @() scope.download(test_ScopeSerial.metadata(1, 0)), ...
                "ScopeSerial:protocol");
            scope.close();

            record = test_ScopeSerial.makeRecord();
            recordText = string(char(record));
            recordText = replace(recordText, "# 1022" + newline, ...
                "## 1022" + newline);
            transport = FakeScopeTransport(uint8(char(recordText)));
            scope = ScopeSerial("fake", Transport=transport);
            testCase.verifyError( ...
                @() scope.download(test_ScopeSerial.metadata(1, 0)), ...
                "ScopeSerial:protocol");
            scope.close();
        end

        function notReadyTimeoutAndLineBoundAreRejected(testCase)
            transport = FakeScopeTransport(uint8(sprintf( ...
                'SCOPE-DATA/1 ERROR NOT_READY ARMED\n')));
            scope = ScopeSerial("fake", Transport=transport);
            testCase.verifyError( ...
                @() scope.download(test_ScopeSerial.metadata(1, 0)), ...
                "ScopeSerial:protocol");
            scope.close();

            transport = FakeScopeTransport(uint8.empty(1, 0));
            scope = ScopeSerial("fake", Transport=transport);
            testCase.verifyError( ...
                @() scope.download( ...
                    test_ScopeSerial.metadata(1, 0), Timeout=0.01), ...
                "ScopeSerial:timeout");
            scope.close();

            transport = FakeScopeTransport( ...
                uint8([repmat('x', 1, 257), newline]));
            scope = ScopeSerial("fake", Transport=transport);
            testCase.verifyError( ...
                @() scope.download(test_ScopeSerial.metadata(1, 0)), ...
                "ScopeSerial:protocol");
            scope.close();
        end

        function metadataValidation(testCase)
            transport = FakeScopeTransport( ...
                test_ScopeSerial.makeRecord());
            scope = ScopeSerial("fake", Transport=transport);
            invalid = {
                struct("decimation", false, "samplePeriodUs", 100, ...
                    "durationMs", 102.4, "pretriggerRatio", 0), ...
                test_ScopeSerial.metadata(0, 0), ...
                test_ScopeSerial.metadata(101, 0), ...
                struct("decimation", 10, "samplePeriodUs", 100, ...
                    "durationMs", 102.4, "pretriggerRatio", 0), ...
                struct("decimation", 1, "samplePeriodUs", 100, ...
                    "durationMs", NaN, "pretriggerRatio", 0), ...
                struct("decimation", 1, "samplePeriodUs", 100, ...
                    "durationMs", 102.4, "pretriggerRatio", Inf)
            };
            for i = 1:numel(invalid)
                metadata = invalid{i};
                testCase.verifyError( ...
                    @() scope.download(metadata), ...
                    "ScopeSerial:validation");
            end
            scope.close();
        end
    end

    methods (Static, Access = private)
        function metadata = metadata(decimation, pretriggerRatio)
            metadata = struct( ...
                "decimation", decimation, ...
                "samplePeriodUs", 100 * decimation, ...
                "durationMs", 102.4 * decimation, ...
                "pretriggerRatio", pretriggerRatio);
        end

        function record = makeRecord(options)
            arguments
                options.FinalIndex = 1022
                options.Channels = ScopeSerial.ChannelNames
                options.ValueCount = 8192
                options.CorruptValue = -1
                options.Terminator = "end record"
            end
            lines = cell(options.ValueCount + 4, 1);
            lines{1} = 'begin record';
            lines{2} = ['#' char(strjoin(options.Channels, ',')) ','];
            lines{3} = sprintf('# %d', options.FinalIndex);
            for index = 0:(options.ValueCount - 1)
                if index == options.CorruptValue
                    lines{index + 4} = 'not-hex!';
                else
                    sample = floor(index / 8);
                    channel = mod(index, 8);
                    bits = typecast(single(sample * 10 + channel), 'uint32');
                    lines{index + 4} = sprintf('%08x', bits);
                end
            end
            lines{end} = char(options.Terminator);
            record = uint8(char(strjoin(string(lines), newline) + newline));
        end
    end
end

File: src/tools/tests/test_scope_serial.py
#
# Copyright (c) 2021-present LAAS-CNRS
#
# SPDX-License-Identifier: GPL-2.0-or-later
#

import struct
import sys
import unittest
from pathlib import Path


TOOLS_DIR = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(TOOLS_DIR))

from scope_serial import (  # noqa: E402
    SCOPE_CHANNEL_NAMES,
    SCOPE_SAMPLE_COUNT,
    ScopeSerial,
    ScopeSerialError,
)


class FakeSerial:
    def __init__(self, probe_response=b"SCOPE-DATA/1 OK\n", record=b""):
        self.probe_response = probe_response
        self.record = record
        self.input = bytearray()
        self.writes = []
        self.closed = False
        self.settings = None

    def factory(self, **settings):
        self.settings = settings
        return self

    def reset_input_buffer(self):
        self.input.clear()

    def write(self, value):
        self.writes.append(value)
        if value == b"?":
            self.input.extend(self.probe_response)
        elif value == b"D":
            self.input.extend(self.record)
        return len(value)

    def flush(self):
        pass

    def read(self, size):
        if not self.input:
            return b""
        result = bytes(self.input[:size])
        del self.input[:size]
        return result

    def close(self):
        self.closed = True


def make_record(
    *,
    final_index=1022,
    channels=SCOPE_CHANNEL_NAMES,
    value_count=SCOPE_SAMPLE_COUNT * len(SCOPE_CHANNEL_NAMES),
    corrupt_value=None,
    terminator="end record",
):
    lines = [
        "begin record",
        "#" + ",".join(channels) + ",",
        f"# {final_index}",
    ]
    for index in range(value_count):
        if corrupt_value is not None and index == corrupt_value:
            lines.append("not-hex!")
        else:
            sample = index // len(SCOPE_CHANNEL_NAMES)
            channel = index % len(SCOPE_CHANNEL_NAMES)
            value = float(sample * 10 + channel)
            lines.append(struct.pack(">f", value).hex())
    lines.append(terminator)
    return ("\n".join(lines) + "\n").encode("ascii")


class ScopeSerialTests(unittest.TestCase):
    def make_scope(self, record, **kwargs):
        fake = FakeSerial(record=record)
        scope = ScopeSerial(
            "fake-scope-port",
            serial_factory=fake.factory,
            **kwargs,
        )
        return scope, fake

    def test_probe_is_explicit_fixed_baud_and_closes_on_mismatch(self):
        with self.assertRaises(ValueError):
            ScopeSerial(None)

        fake = FakeSerial(probe_response=b"WRONG PORT\n")
        with self.assertRaises(ScopeSerialError):
            ScopeSerial(
                "fake-scope-port",
                probe_timeout=0.001,
                serial_factory=fake.factory,
            )
        self.assertTrue(fake.closed)
        self.assertEqual(fake.settings["baudrate"], 115200)

    def test_valid_capture_is_decoded_and_rotated(self):
        scope, fake = self.make_scope(make_record())
        capture = scope.download(
            decimation=10,
            sample_period_us=1000,
            duration_ms=1024.0,
            pretrigger_ratio=0.2,
        )

        self.assertEqual(fake.writes, [b"?", b"D"])
        self.assertEqual(capture.channel_names, SCOPE_CHANNEL_NAMES)
        self.assertEqual(capture.final_index, 1022)
        self.assertEqual(capture.samples[0][0], 10230.0)
        self.assertEqual(capture.samples[1][0], 0.0)
        self.assertEqual(len(capture.samples), SCOPE_SAMPLE_COUNT)
        self.assertEqual(len(capture.samples[0]), len(SCOPE_CHANNEL_NAMES))
        self.assertAlmostEqual(capture.time_axis_s[0], -0.2048)
        scope.close()
        self.assertTrue(fake.closed)

    def test_truncated_nonhex_and_wrong_count_payloads_are_rejected(self):
        cases = (
            ("truncated", make_record(value_count=8191), "truncated"),
            (
                "nonhex",
                make_record(corrupt_value=123),
                "not 8-digit hexadecimal",
            ),
            (
                "extra",
                make_record(value_count=8193),
                "more than 8192 values",
            ),
        )
        for name, record, message in cases:
            with self.subTest(name=name):
                scope, _ = self.make_scope(record)
                with self.assertRaisesRegex(ScopeSerialError, message):
                    scope.download(
                        decimation=1,
                        sample_period_us=100,
                        duration_ms=102.4,
                        pretrigger_ratio=0.0,
                    )

    def test_wrong_channel_and_malformed_final_index_are_rejected(self):
        wrong_channels = ("wrong",) + SCOPE_CHANNEL_NAMES[1:]
        scope, _ = self.make_scope(make_record(channels=wrong_channels))
        with self.assertRaisesRegex(ScopeSerialError, "header mismatch"):
            scope.download(
                decimation=1,
                sample_period_us=100,
                duration_ms=102.4,
                pretrigger_ratio=0.0,
            )

        record = make_record().replace(b"# 1022\n", b"## 1022\n", 1)
        scope, _ = self.make_scope(record)
        with self.assertRaisesRegex(ScopeSerialError, "final-index"):
            scope.download(
                decimation=1,
                sample_period_us=100,
                duration_ms=102.4,
                pretrigger_ratio=0.0,
            )

    def test_not_ready_timeout_and_bounded_line_are_rejected(self):
        scope, _ = self.make_scope(
            b"SCOPE-DATA/1 ERROR NOT_READY ARMED\n"
        )
        with self.assertRaisesRegex(ScopeSerialError, "NOT_READY ARMED"):
            scope.download(
                decimation=1,
                sample_period_us=100,
                duration_ms=102.4,
                pretrigger_ratio=0.0,
            )

        scope, _ = self.make_scope(b"")
        with self.assertRaisesRegex(ScopeSerialError, "timed out"):
            scope.download(
                decimation=1,
                sample_period_us=100,
                duration_ms=102.4,
                pretrigger_ratio=0.0,
                timeout=0.001,
            )

        scope, _ = self.make_scope(b"x" * 257 + b"\n")
        with self.assertRaisesRegex(ScopeSerialError, "exceeds 256"):
            scope.download(
                decimation=1,
                sample_period_us=100,
                duration_ms=102.4,
                pretrigger_ratio=0.0,
            )

    def test_capture_metadata_validation(self):
        scope, _ = self.make_scope(make_record())
        invalid = (
            dict(
                decimation=True,
                sample_period_us=100,
                duration_ms=102.4,
                pretrigger_ratio=0.0,
            ),
            dict(
                decimation=0,
                sample_period_us=100,
                duration_ms=102.4,
                pretrigger_ratio=0.0,
            ),
            dict(
                decimation=101,
                sample_period_us=10100,
                duration_ms=10342.4,
                pretrigger_ratio=0.0,
            ),
            dict(
                decimation=10,
                sample_period_us=100,
                duration_ms=102.4,
                pretrigger_ratio=0.0,
            ),
            dict(
                decimation=1,
                sample_period_us=100,
                duration_ms=float("nan"),
                pretrigger_ratio=0.0,
            ),
            dict(
                decimation=1,
                sample_period_us=100,
                duration_ms=102.4,
                pretrigger_ratio=float("inf"),
            ),
        )
        for metadata in invalid:
            with self.subTest(metadata=metadata), self.assertRaises(ValueError):
                scope.download(**metadata)


if __name__ == "__main__":
    unittest.main()
